### PR TITLE
[ao][sparsity] Base class for Data Sparsifiers

### PR DESCRIPTION
**Pull request by macandro96**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #79141

Base Data Sparsifier class for all Data sparsifiers.
The abstract class accepts raw torch tensors / embedding / embedding bags (refer to SUPPORTED_TYPES above)
to prepare for sparsification.
In this case, mask (and parametrizations) is owned by the class and not by the user.
Specifically, the container object inside the class maintains the mask and parametrizations of the input data

Abstract methods that need to be implemented:
- update_mask: Function to compute a new mask for all keys in the
    `data_groups`.
     Should contain at least 2 arguments - name, data.

Test Plan:
```python test/test_ao_sparsity.py TestBaseDataSparsifier```

### CODE REVIEW

File: torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+                        'container_state_dict': self._container.state_dict()})
+
+    def squash_mask(self, *args, **kwargs):
+        r"""Squashes the sparse masks into the appropriate tensors.
+        """
+        for name in self.data_groups.keys():
+            parametrize.remove_parametrizations(self._container, name, leave_parametrized=True)
+       
+    def convert(self):
+        # TODO: Call the torch.ao.utils.convert in here
```
Comment (z-a-f) [inline torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py:227]: Remove this line

**Review by z-a-f — CHANGES_REQUESTED** [review]
It seems to me there are >1 different things being contributed here:

1. The parametrizations container -- let's split it into a separate diff + add tests if that would work from the point of view of access
1. The support for embeddings and tensors -- that seems like two different functionalities, and, IMHO, should be contributed individually (basically split the diff with one PR taking care of embeddings and another taking care of tensors
1. BaseSparsifier -- this could be inherited from the original BaseSparsifier, with some overrides of the internal variables and functionality. That way you can reuse the `__state_dict__` utilities

File: torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+                data = self.get_data(name, return_sparsified=False)
+                # need name for the mask otherwise can directly pass mask?
+                self.update_mask(name, data, **config)
+
+    @abc.abstractmethod
+    def update_mask(self, name, data, **kwargs):
+        pass
+
+
+__all__ = [Dict, Any, FakeSparsity, List, Optional, Tuple, defaultdict, 'BaseDataSparsifier', 'Container']
```
Comment (z-a-f) [inline torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py:246]: `__all__` should be inside the `torch/ao/sparsity/experimental/data_sparsifier/__init__.py`

File: torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+        return self.state[name]['mask']
+
+    def __setstate__(self, state):
+        self.__dict__.update(state)
+
+    def __repr__(self):
+        format_string = self.__class__.__name__ + ' ('
+        for name, sparse_args in self.data_groups.items():
+            format_string += '\n'
+            format_string += f'\Data Group\n'
```
Comment (z-a-f) [inline torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py:141]: Careful with escape chars

File: torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+        else:
+            return getattr(self._container, name)
+
+    def get_mask(self, name):
+        if name not in self.state:
+            raise ValueError("data with specified name does not exist")
+        return self.state[name]['mask']
+
+    def __setstate__(self, state):
+        self.__dict__.update(state)
```
Comment (z-a-f) [inline torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py:135]: The `self._container` is not loaded here. Consider the following:

```python
def __setstate__(self, state):
    container_state = state.pop('container_state')
    self.__dict__.update(state)
    self._container.load_from_state_dict(container_state)
```

**Comment by macandro96** [comment]
Created a new one [here](https://github.com/pytorch/pytorch/pull/79254)
